### PR TITLE
fixes and improvements

### PR DESCRIPTION
**Pull request by mahantesh-ais**
OXT-1221

### CODE REVIEW

File: control/tap-ctl.c [hunk 1]
```diff
@@ -399,8 +406,39 @@ tap_cli_destroy(int argc, char **argv)
...
+			return -EINVAL;
+		}
+
+		minor = minor(sb.st_rdev);
+	}
+
+	if (minor == -1)
+ 		goto usage;
+
+	if (pid == -1) {
```
Comment (eric-ch) [inline control/tap-ctl.c:459]: `atoi(3)` returns 0 on invalid strings. This being used everywhere else, it should probably be changed in a different PR anyway.

File: control/tap-ctl.c [hunk 2]
```diff
@@ -399,8 +406,39 @@ tap_cli_destroy(int argc, char **argv)
...
+		pid = tap_ctl_find_pid(minor);
+		if (pid == -1) {
```
Comment (eric-ch) [inline control/tap-ctl.c]: `tap_ctl_find_pid()` defined later in this PR can return `-ENOENT` (-2), or other `-errno` values from nested functions. This test is invalid (that snippet is repeated a lot).

File: drivers/Makefile.am [hunk 3]
```diff
@@ -1,19 +1,21 @@
+SUBDIRS = crypto
 AM_CFLAGS  = -Wall
-AM_CFLAGS += -Werror
+AM_CFLAGS += -Werror -Wno-unused-function
```
Comment (eric-ch) [inline drivers/Makefile.am]: It would be better to remove them or add a condition on macro that can be defined at configuration or build time instead.

File: drivers/block-crypto.c [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * Copyright (c) 2014 Citrix Systems, Inc.
+ */
```
Comment (eric-ch) [inline drivers/block-crypto.c:31]: This header seems mal-formatted.

File: drivers/block-crypto.c [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+	list_add(&ent->next, &keystore);
+
+out:
+	if (err)
+		keystore_free_entry(ent);
+	return err;
+}
+
+static int
+keystore_read_entry(uint8_t *keybuf, int *keysize, const struct vhd_keyhash *keyhash)
```
Comment (eric-ch) [inline drivers/block-crypto.c]: `keystore_*` functions appear to be unused.

File: tapback/backend.c [hunk 6]
```diff
@@ -742,7 +742,7 @@ tapback_backend_probe_device(backend_t *backend,
      * We don't set a XenStore watch on these paths in order to limit the
      * number of watches for performance reasons.
      */
-    if (!remove && comp) {
+    if (!remove && comp && device) {
```
Comment (eric-ch) [inline tapback/backend.c:745]: I do not see a path how `&& device` would be false with `!remove` being checked before.

File: tapback/frontend.c [hunk 7]
```diff
@@ -440,20 +440,15 @@ frontend_changed(vbd_t * const device, const XenbusState state)
 
     switch (state) {
         case XenbusStateInitialising:
-			if (device->hotplug_status_connected)
 				err = xenbus_switch_state(device, XenbusStateInitWait);
```
Comment (eric-ch) [inline tapback/frontend.c]: Indentation.
Also, why does this need to be removed?
```
/local/domain/0/backend/vbd/<domid>/<vbdid>/hotplug-status = "connected"
```
Seems to be populated, is it happening too late?

File: tapback/frontend.c [hunk 8]
```diff
@@ -509,30 +504,7 @@ tapback_backend_handle_otherend_watch(backend_t *backend,
...
-            err = 0;
-            if (!xs_rm(device->backend->xs, XBT_NULL, _path)) {
-                if (errno != ENOENT) {
-                    err = errno;
-                    WARN(device, "failed to remove %s: %s\n", path,
-                            strerror(err));
-                }
-            }
-		}
-    } else {
```
Comment (eric-ch) [inline tapback/frontend.c:535]: Indentation (from original code it seems, mixing tabs and spaces).

That block looks ok though: If the backend no longer finds the frontend, should it not remove its Xenstore entries?

**Review by eric-ch — CHANGES_REQUESTED** [review]

File: control/tap-ctl.c (hunk 1 above)
Comment (mahantesh-ais) [reply control/tap-ctl.c:459]: Ack, will fix it here. But this is legacy code from blktap2 patch and needs to be fixed in OXT version/branch (if any) using blktap2.

File: control/tap-ctl.c (hunk 2 above)
Comment (mahantesh-ais) [reply control/tap-ctl.c]: Ack, will fix it here. But this is legacy code from blktap2 patch and needs to be fixed in OXT version/branch (if any) using blktap2.

File: drivers/Makefile.am (hunk 3 above)
Comment (mahantesh-ais) [reply drivers/Makefile.am]: Patches ported from blktap2 had some unused functions, I retained them being unaware of their usage. As suggested by you, I will encapsulate these functions under a macro.

File: drivers/block-crypto.c (hunk 5 above)
Comment (mahantesh-ais) [reply drivers/block-crypto.c]: Patches ported from blktap2 had some unused functions, I retained them being unaware of their usage. As suggested by you, I will encapsulate these functions under a macro.

File: tapback/backend.c (hunk 6 above)
Comment (mahantesh-ais) [reply tapback/backend.c:745]: tapback_backend_find_device is actually a macro. This code path is triggered during iso hotswap when the device is NULL for a short amount of time.

File: tapback/frontend.c (hunk 7 above)
Comment (mahantesh-ais) [reply tapback/frontend.c]: Yes, hotplug-status is populated after the frontend state is changed to connected (XenbusStateConnected)

File: tapback/frontend.c (hunk 8 above)
Comment (mahantesh-ais) [reply tapback/frontend.c:535]: Ack, will fix the indentation.
That block of code removing the Xenstore entries is never triggered and hence redundant.

File: drivers/Makefile.am (hunk 3 above)
Comment (mahantesh-ais) [reply drivers/Makefile.am]: After going through unused functions decided to remove them.

File: drivers/block-crypto.c (hunk 5 above)
Comment (mahantesh-ais) [reply drivers/block-crypto.c]: After going through unused functions decided to remove them.

File: drivers/block-crypto.c (hunk 4 above)
Comment (mahantesh-ais) [reply drivers/block-crypto.c:31]: This comes from legacy code, I'm not sure what is the correct format. Please let me know what is the correct format if this needs to be corrected.

**Comment by mahantesh-ais** [comment]
Fixed.

File: drivers/block-crypto.c (hunk 4 above)
Comment (eric-ch) [reply drivers/block-crypto.c:31]: Actually I was told this should be left as it is, my bad.

File: control/tap-ctl.c (hunk 1 above)
Comment (eric-ch) [reply control/tap-ctl.c:459]: Sure, thanks for fixing it in this version.

File: tapback/backend.c (hunk 6 above)
Comment (eric-ch) [reply tapback/backend.c:745]: I got it now, all good, thanks.

**Comment by jean-edouard** [comment]
LGTM, merging around end of day.
